### DISCUSSION TITLE
How do I refresh creds of existing pgxpool?

### DISCUSSION THREAD

**Discussion by vrv7567** [jackc/pgx#2103] [Q&A] [answered]
If we have an existing connection pool, that needs to be refreshed with new credentials, (assume old credentials are still valid since we want a graceful switchover) how can I refresh creds of an existing connection pool?

I'm thinking I should be setting `BeforeAcquire` from `pgxpool.Config` to invalidate existing connections when creds are refreshed. This should close all the connections.

l should also set `BeforeConnect` to ensure I fetch latest set of creds from some external source like secretsmanager? But the disadvantage here being if connections are closed because of their max lifetime expiry eventhough creds haven't changed, we would make unnecessary requests to this external secret-store

What other well-established strategies is the community already using? Are there better steps for my use-case

**Comment by vrv7567** [comment]
or one more strategy would be to have some kind of notification mechanism where if creds are refreshed, 
1. mutex lock is acquired 
2. Wait until all connections in the pool are idle. If yes, close the pool
3. Create a new pool with new creds and assign it variable previously holding the pool
4. unlock mutex

**Comment by jackc** [comment, marked as answer]
I'm not sure about your exact use case, but `BeforeConnect` is the correct place to mutate the connection parameters when making a connection.

As far as closing old connections, you probably want to call `Pool.Reset`.
